Add material on branches (#13)
empa-scientific-it/how-to-git: tutorial.md

### Branching out topics

Often, you will want to work on different topics in parallel and keep them separate from the main codebase.
Those include the development of new features, bug fixes, or experiments.
Git is offering the possibility for branching out.
This is like working on isolated copies of your files, try out new things until everything works: in Git, this is called a branch.
After the person is satisfied with the changes, they can merge them back into the main codebase (`main` branch or `master` branch).

The main branch in Git is often called `main` or `master`, and it represents the stable version of your code.
When you want to work on a new feature or fix a bug, you create a new branch from the `main` branch.
One should not hesitate to create branches, cause they are lightweight and simple to create and merge.

To create a new branch, you can use the GitHub interface.

> There are different ways how to create branches: From the repository; In the dialogue window for a commit; from an issue - and probably there are even more possibilities.

Author: done in https://github.com/empa-scientific-it/how-to-git/pull/13/commits/e26b39b01528112d14c6f9c4ac631de5bd36e21d

You can now make changes to the files.
Once you commit them, they will be pushed, as in saved, to that branch that you chose from the drop-down in the beginning.

### Merge branches

Once you are happy with your work done in a branch, you can merge it back into the `main` branch.
This ensures that your changes are integrated into the main codebase, so they can be used by others.
To do that, you can use the GitHub interface:
1. Go to your repository.
2. Click on the "Pull requests" tab.
3. Click on the green "New pull request" button.

![New pull request button example](images/new-pull-request.png)

4. Select the branch you want to merge from (e.g., `feature-xyz`) and the branch you want to merge into (e.g., `main`).
   After that, you will see a comparison of the two branches and a list of commits that will be merged.
5. Click on the green "Create pull request" button.
6. In the newly opened page, you can add a title and a description for your pull request.
Later you will be able to find that pull request by that title.
   This is a good place to explain what changes you made and why they are important.

>    Also, you can assign reviewers to your pull request, so they can review your changes and give feedback.

Reviewer: maybe later we can link this to the not yet existing part about reviews?

7. Once you are done, click on the green "Create pull request" button again.

After that, your pull request will be created and you can see it in the "Pull requests" tab.
You can continue to work on your branch, and after every commit, the pull request will be updated with the new changes.
Once the work is done, you can merge the pull request by clicking on the green button that says "Merge pull request", "Squash and merge", or "Rebase and merge".

> Usually, you can use Merge or Squash, but more on this topic can be found [here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges).

Reviewer: ```suggestion
Usually, you use Merge or Squash, more details on this topic can be found [here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges).
```

### Branching out topics

Often, you will want to work on different topics in parallel and keep them separate from the main codebase.

> Those include new features, bug fixes, or experiments.

Reviewer: ```suggestion
Those includes the development of new features, bug fixes, or experiments.
```

This is where Git's branching capabilities come into play.

The main branch in Git is often called `main` or `master`, and it represents the stable version of your code.
When you want to work on a new feature or fix a bug, you create a new branch from the `main` branch.

> One should not hesitate to create branches, cause they are cheap to create and merge.

Author: fixed in https://github.com/empa-scientific-it/how-to-git/pull/13/commits/07e9cd5defec690468406fb1510ddc738f5fb25e

> To create a new branch, you can use the GitHub interface:

Reviewer: ```suggestion
To create a new branch, you can use the GitHub interface.
There are different ways how to create branches: From the repository; In the dialogue window for a commit; from an issue - and probably there are even more possibilities.
Following the one from the repository is explained:
```

1. Go to your repository.
2. Click on the "main" dropdown button.

![Branch dropdown example](images/branch-dropdown.png)

3. In the text field, type the name of your new branch (e.g., `feature-xyz`).
4. Press Enter to create the branch.

> And now you can switch to this branch and start working on your new feature or bug fix.

Author: done in https://github.com/empa-scientific-it/how-to-git/pull/13/commits/01876a707a90316ff074b3e2f2644753a7683d81

> If you are curious what it looks like in the commit history, you can go to "Insights" tab of your repository and select "Network" from the left sidebar.

Reviewer: point three in [here](#check-the-history)

I already explained a bit about that, maybe we can shorten the explanation here a bit? or at least also point to that section, or we just leave it as it is?

Author: link is fixed in https://github.com/empa-scientific-it/how-to-git/pull/13/commits/992b1ab5377e5030eeb52a62479f84db81524008

This will show you a graphical representation of your branches and commits.
You will notice that the new branch is simply a pointer to a specific commit in the history.
If the branch is freshly created, it will point to the same commit as the `main` branch, so the code in the new branch is still the same as in the `main` branch.
However, as you make changes in the new branch, it will diverge from the `main` branch and will have parts that are different from the `main` branch.

> ### Switch between topics (branches)

Reviewer: ```suggestion
### Switch between branched out topic
```

Reviewer: *topics

To switch between branches, you can use the GitHub interface:
1. Go to your repository.
2. Click on the "main" dropdown button right below the repository name.

![Branch dropdown example](images/branch-dropdown.png)

> 3. Select the branch you want to switch to (e.g., `feature-xyz`).

Reviewer: ```suggestion
3. Select the branch for your topic of interest you want to switch to (e.g., `feature-xyz`).
```

> Once you switch to a branch, the files in your repository will be updated to reflect the state of that branch.

Reviewer: ```suggestion
Once you switch to a branch, the files in your repository will be updated to reflect the state of that branch.
Meaning, it can be that it looks different from before, since you changed files for new features or bug fixes.
```

> You can now make changes to the files, commit them, and push them to the selected branch.

Reviewer: ```suggestion
You can now make changes to the files.
Once you commit them, they will be pushed, as in saved, to that branch that you chose from the drop down in the beginning
```

> ### Develop in a branch

Reviewer: Somehow I feel the title is not enough self explanatory.
Maybe something like "Developing new features in a branch"?

Reviewer: After reading the whole chapter here, I feel most of it is already explained above in the chapters.
I would suggest, instead of having a whole chapter about that, in the top somewhere a introduction say, developing new things in branches is a common practise and leave out lines 112 to 121

Reviewer: I put a comment on line 75 for that.

Author: Sure, I will drop that paragraph.

Author: done in https://github.com/empa-scientific-it/how-to-git/pull/13/commits/a0aa4a7b6dbff261ab958ac35aa9fd1b4fe49d13

Developing in a branch is a common practice in Git.
It allows you to work on new features or bug fixes without affecting the main codebase.

That also gives you a freedom to experiment with your code, add changes, remove changes and even rewrite the history of your branch.
The idea here is that once your changes are ready, you can merge them back into the `main` branch and write a clean commit message that describes the changes you made.
Once you are in a branch, you can make changes to your files, commit them and push them to the remote repository.
You can also create new files, delete files, and rename files - everything will remain in the branch until you decide to merge it back into the `main` branch.

### Merge branches

> Once you are done with your work in a branch, you can merge it back into the `main` branch making sure that your changes are integrated into the main codebase.

Reviewer: ```suggestion
Once you are happy with your new feature of bug fix, you merge it to the `main` branch.
Like that you ensure that your changes will be back to the original file, represented in the stable version of your code.
```

Author: done in https://github.com/empa-scientific-it/how-to-git/pull/13/commits/3f1272825b615b412f0eaed359f7eb5ab01e5dc9

To do that, you can use the GitHub interface:
1. Go to your repository.
2. Click on the "Pull requests" tab.
3. Click on the green "New pull request" button.

![New pull request button example](images/new-pull-request.png)

4. Select the branch you want to merge from (e.g., `feature-xyz`) and the branch you want to merge into (e.g., `main`).
   After that, you will see a comparison of the two branches and a list of commits that will be merged.
5. Click on the green "Create pull request" button.

> 6. In the newly opened page, you can add a title and a description for your pull request.

Reviewer: ```suggestion
6. In the newly opened page, you can add a title and a description for your pull request.
Later you will be able to find that pull request by that title.
```

   This is a good place to explain what changes you made and why they are important.
   Also, you can assign reviewers to your pull request, so they can review your changes and give feedback.
7. Once you are done, click on the green "Create pull request" button again.

After that, your pull request will be created and you can see it in the "Pull requests" tab.
You can continue to work on your branch, and after every commit, the pull request will be updated with the new changes.

> Once the work is done, you can merge the pull request by clicking on the green button that says "Merge pull request", or "Squash and merge", or "Rebase and merge".

Reviewer: ```suggestion
Once the work is done, you can merge the pull request by clicking on the green button that says "Merge pull request", or "Squash and merge", or "Rebase and merge".
Usually you can use Merge or Squash, but more to this topic can be found [here](https://docs.github.com/de/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges)
```
